libass/ass_utils.c

```diff
@@ -31,41 +31,6 @@
 #include "ass_utils.h"
 #include "ass_string.h"
 
-#if (defined(__i386__) || defined(__x86_64__)) && CONFIG_ASM
-
-#include "x86/cpuid.h"
-
-int has_sse2(void)
-{
-    uint32_t eax = 1, ebx, ecx, edx;
-    ass_get_cpuid(&eax, &ebx, &ecx, &edx);
-    return (edx >> 26) & 0x1;
-}
-
-int has_avx(void)
-{
-    uint32_t eax = 1, ebx, ecx, edx;
-    ass_get_cpuid(&eax, &ebx, &ecx, &edx);
-    if (!(ecx & (1 << 27))) // not OSXSAVE
-        return 0;
-    uint32_t misc = ecx;
-    ass_get_xgetbv(0, &eax, &edx);
-    if ((eax & 0x6) != 0x6)
-        return 0;
-    eax = 0;
-    ass_get_cpuid(&eax, &ebx, &ecx, &edx);
-    return (ecx & 0x6) == 0x6 ? (misc >> 28) & 0x1 : 0; // check high bits are relevant, then AVX support
```

Comment:
There seems to be no equivalent of the `(ecx & 0x6) == 0x6` in the replacement code. However afaict the new version is correct and the old incorrect. On my Zen+-CPU, the old `has_…` functions actually claim it would only supported SSE2, while the new `ass_get_cpu_flags` correctly identifies SSE2, AVX and AVX2 to all be supported. Does this also affect other µarches?
The `ass_get_cpu_flags` commit-message should document the fix.

Reply:
First and third commits have taken verbatim from @rcombs [checkasm](https://github.com/libass/libass/pull/400/commits) branch (third had merge conflict but only with context and not actual changeset) so I've not looked deeply into them. I'll dive deeper when I have time.

Reply:
Actually this CPU-flag business is orthogonal to moving code to separate directory. So I should postpone it to an actual checkasm PR.

Reply:
Just my 2¢, perhaps irrelevant: I remember x86 CPUID checks being tricky in the sense that some extensions may be supported by the CPU but not by the OS, in which case they mustn’t be used. Could it be that the 0x6 mask checked something along these lines? At any rate, whenever anyone gets around to re-checking this, they need to be _careful._

Reply:
I tested this on some more CPUs, the results are that the new (now removed) version was correct in all tests, the old version wrong for all AMD CPUs. In more detail the old version reports (putting what of the tested the architecture actually supports in paranthises):
 - AMD Zen+ (SSE2, AVX, AVX2): **SSE2**
 - AMD Puma+ (SSE2, AVX): **SSE2**
 - Intel Kaby Lake (SSE2, AVX, AVX2): **SSE2, AVX, AVX2**
 
 ---
 
 [Reading up](https://en.wikipedia.org/w/index.php?title=CPUID&oldid=1040897815#EAX=0:_Highest_Function_Parameter_and_Manufacturer_ID) on what info CPUID-leaf 0 contains, `ebx`, `edx` and `ecx` simply contains a 12byte SCII-string of the manufacturer name. The `(ecx & 0x6) == 0x6` is a spot-check which is true for `"GenuineINtel"` and false for `"AuthenticAMD"`. Thus preventing all AMD, Via, Cyrix, etc and also many Hypervisors, VMs and emulators — *even if running on a physical Intel CPU* — from making use of AVX and AVX2.
 As stated before, I'm no expert in x86(-assembly), but this seems undesireable to me. At least the CPUID-page of the English Wikipedia does not note any difference in the encoding of any of the feature flags we check between different manufacturers.

Reply:
> […] So I should postpone it to an actual checkasm PR.

Since this blocks non-Intel- and virtual-CPUs from making use of our AVX2-routines and after discussion with rcombs on IRC, there appears to be no need for such a (broad) block, I'd prefer to have this fixed soon'ish, though not necessarily (but possibly) in this PR.
Whether to do this independent or together (but documented) with the BitmapEngine-change, I have no opinion on.
Same goes for whether to check for and block some of the very first AVX (but iinm not AVX2) CPUs by AMD like [FFmpeg partially does](https://git.ffmpeg.org/gitweb/ffmpeg.git/blob/1ad4782d2a17d9fed38c0356f1c9f790973eec01:/libavutil/x86/cpu.c#l191) (those *may be* slower with AVX than SSE2).